examples/neighbors/approximate_nearest_neighbors.py

```diff
@@ -0,0 +1,290 @@
+        self.method = method
+        self.metric = metric
+        self.n_jobs = n_jobs
+
+    def fit(self, X):
+        self.n_samples_fit = X.shape[0]
+
+        # see more metric in the manual
+        # https://github.com/nmslib/nmslib/tree/master/manual
+        space = {
+            'sqeuclidean': 'l2',
+            'euclidean': 'l2',
+            'cosine': 'cosinesimil',
+            'l1': 'l1',
+            'l2': 'l2',
+        }[self.metric]
+
+        self.nmslib_ = nmslib.init(method=self.method, space=space)
+        self.nmslib_.addDataPointBatch(X)
+        self.nmslib_.createIndex()
+        return self
+
+    def transform(self, X):
+        n_samples_transform = X.shape[0]
+
+        # For compatibility reasons, as each sample is considered as its own
+        # neighbor, one extra neighbor will be computed.
+        n_neighbors = self.n_neighbors + 1
+
+        results = self.nmslib_.knnQueryBatch(X, k=n_neighbors,
+                                             num_threads=self.n_jobs)
+        indices, distances = zip(*results)
+        indices, distances = np.vstack(indices), np.vstack(distances)
+
+        if self.metric == 'sqeuclidean':
+            distances **= 2
+
+        indptr = np.arange(0, n_samples_transform * n_neighbors + 1,
+                           n_neighbors)
+        kneighbors_graph = csr_matrix((distances.ravel(), indices.ravel(),
+                                       indptr), shape=(n_samples_transform,
+                                                       self.n_samples_fit))
+
+        return kneighbors_graph
+
+    def fit_transform(self, X, y=None):
+        return self.fit(X).transform(X)
+
+
+class AnnoyTransformer(BaseEstimator, TransformerMixin):
+    """Wrapper for using annoy.AnnoyIndex as sklearn's KNeighborsTransformer"""
+
+    def __init__(self, n_neighbors=5, metric='euclidean', n_trees=10,
+                 search_k=-1):
+        self.n_neighbors = n_neighbors
+        self.n_trees = n_trees
+        self.search_k = search_k
+        self.metric = metric
+
+    def fit(self, X):
+        self.n_samples_fit = X.shape[0]
+        metric = self.metric if self.metric != 'sqeuclidean' else 'euclidean'
+        self.annoy_ = annoy.AnnoyIndex(X.shape[1], metric=metric)
+        for i, x in enumerate(X):
+            self.annoy_.add_item(i, x.tolist())
+        self.annoy_.build(self.n_trees)
+        return self
+
+    def transform(self, X):
+        if X is None:
+            n_samples_transform = self.n_samples_fit
+        else:
+            n_samples_transform = X.shape[0]
+
+        # For compatibility reasons, as each sample is considered as its own
+        # neighbor, one extra neighbor will be computed.
+        n_neighbors = self.n_neighbors + 1
+
+        indices = np.empty((n_samples_transform, n_neighbors),
+                           dtype=np.int)
+        distances = np.empty((n_samples_transform, n_neighbors))
+
+        if X is None:
```

Comment:
I do not think we have used `X=None` as a pattern to go down a different path in `transform`. It would be weird to showcase a pattern we do not use internally. Maybe something like this:

```py
def transform(self, X):
    return self._transform(X)

def fit_transform(self, X):
    return self.fit(X)._transform(None)

def _transform(self, X):
    n_samples_transform = self.n_samples_fit_ if X is None else X.shape[0]
    ...
	return kneighbors_graph
```